gusto/initialisation_tools.py

```diff
@@ -5,10 +5,70 @@
 
 from __future__ import absolute_import
 from firedrake import MixedFunctionSpace, TrialFunctions, TestFunctions, \
+    TestFunction, TrialFunction, \
     FacetNormal, inner, div, dx, ds_b, ds_t, DirichletBC, \
     Expression, Function, Constant, \
     LinearVariationalProblem, LinearVariationalSolver, \
-    NonlinearVariationalProblem, NonlinearVariationalSolver, split
+    NonlinearVariationalProblem, NonlinearVariationalSolver, split, solve
+
+
+def incompressible_hydrostatic_balance(state, b0, p0, top=False, params=None):
+
+    # get F
+    v = TrialFunction(state.Vv)
+    w = TestFunction(state.Vv)
+
+    unp1 = state.xnp1.split()[0]
+    dim = unp1.ufl_element().value_shape()[0]
+    bc = ("0.0",)*dim
+
+    if top:
+        bstring = "top"
+    else:
+        bstring = "bottom"
+
+    bcs = [DirichletBC(state.Vv, Expression(bc), bstring)]
+
+    a = inner(w, v)*dx
+    L = inner(state.k, w)*b0*dx
+    F = Function(state.Vv)
+
+    solve(a == L, F, bcs=bcs)
+
+    # define mixed function space
+    WV = (state.Vv)*(state.V[1])
+
+    # get pprime
+    v, pprime = TrialFunctions(WV)
+    w, phi = TestFunctions(WV)
+
+    bcs = [DirichletBC(WV[0], Expression(bc), bstring)]
+
+    a = (
+        inner(w, v) + div(w)*pprime + div(v)*phi
+    )*dx
+    L = phi*div(F)*dx
+    w1 = Function(WV)
+
+    solver_parameters = {'ksp_type':'gmres',
+                         'pc_type': 'fieldsplit',
+                         'pc_fieldsplit_type': 'schur',
+                         'pc_fieldsplit_schur_fact_type': 'full',
+                         'pc_fieldsplit_schur_precondition': 'selfp',
+                         'fieldsplit_1_ksp_type': 'preonly',
+                         'fieldsplit_1_pc_type': 'gamg',
+                         'fieldsplit_1_mg_levels_pc_type': 'bjacobi',
+                         'fieldsplit_1_mg_levels_sub_pc_type': 'ilu',
+                         'fieldsplit_0_ksp_type': 'richardson',
+                         'fieldsplit_0_ksp_max_it': 4,
+                         'ksp_atol': 1.e-08,
```

Comment:
I don't think atol of 1e-8 is necessary. It is one of legacy settings in the old code.